### PR TITLE
Add ORM EntityFramework and implement Repository pattern

### PR DESCRIPTION
**Pull request by mtarcha**
Add EntityFramework sdk.
Create Repository.
Create RepositoeyFactory.
Add DataContracts: User,UserProfile, Friends.
Add settings to server.console project.

@iivchenko-softheme

### CODE REVIEW

**Comment by iivchenko** [comment]
- [x] Fix grammar in the header of the pull request.

**Comment by iivchenko** [comment]
- [x] Check headers please

**Comment by iivchenko** [comment]
:exclamation: 
Wrong repository. Please use you fork instead of origin for branches and pull requests.

File: Common/Common.DataBaseRepositories/IRepositoryFactory.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿// // <copyright company="Tarcha & Ivchenko Company">
+// //      Copyright (c) 2015, All Right Reserved
+// // </copyright>
+// // <author>Ivan Ivchenko</author>
+// // <author>Myroslava Tarcha</author>
+
+using Common.DataContracts;
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/IRepositoryFactory.cs]: - [x] Unused

File: Common/Common.DataBaseRepositories/IRepository.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq.Expressions;
+
+namespace DataBaseRepositories
+{
+    /// <summary>
+    /// Provides access to entities in DB.
+    /// </summary>
+    /// <typeparam name="TEntity">The type of elements that are/have to be presented in DB.</typeparam>
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/IRepository.cs]: - [x] not obvious

**Comment by iivchenko** [comment]
- [x] lets check xml documentation togather

File: Common/Common.DataBaseRepositories/IRepository.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    public interface IRepository<TEntity> where TEntity : class
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/IRepository.cs]: - [ ] why can't we use valuable types?

**Comment by iivchenko** [comment]
- [x] lets check justifications derogate

File: Common/Common.DataBaseRepositories/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+// <author>Myroslava Tarcha</author>
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+// Управление общими сведениями о сборке осуществляется с помощью 
+// набора атрибутов. Измените значения этих атрибутов, чтобы изменить сведения,
+// связанные со сборкой.
+[assembly: AssemblyTitle("DataBaseRepositories")]
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Properties/AssemblyInfo.cs]: - [x] assembly. go up

File: Common/Common.DataBaseRepositories/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+// </copyright>
+// <author>Ivan Ivchenko</author>
+// <author>Myroslava Tarcha</author>
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+// Управление общими сведениями о сборке осуществляется с помощью 
+// набора атрибутов. Измените значения этих атрибутов, чтобы изменить сведения,
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Properties/AssemblyInfo.cs]: - [x] look attentive through the file

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 6]
```diff
@@ -0,0 +1,281 @@
+﻿// // <copyright company="Tarcha & Ivchenko Company">
+// //      Copyright (c) 2015, All Right Reserved
+// // </copyright>
+// // <author>Ivan Ivchenko</author>
+// // <author>Myroslava Tarcha</author>
+
+using System;
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs:7]: - [x] Check namespaces

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+using NLog;
+
+namespace DataBaseRepositories
+{
+    /// <summary>
+    /// Provides base methods for working with entities.
+    /// </summary>
+    /// <typeparam name="TEntity">The mapped type.</typeparam>
+    /// <typeparam name="TContext">Type of the used DbContext.</typeparam>
+    public class Repository<TEntity, TContext> : IRepository<TEntity>, IDisposable
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [x] It is better to do IRepository as IDisposable

Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [x] Lets mark the class as sealed, delete Logger property and DataContext property. Lets logger name will be repository+entity

File: Common/Common.DataBaseRepositories/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+[assembly: AssemblyTitle("DataBaseRepositories")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Microsoft")]
+[assembly: AssemblyProduct("DataBaseRepositories")]
+[assembly: AssemblyCopyright("Copyright © Microsoft 2015")]
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
+
+// Параметр ComVisible со значением FALSE делает типы в сборке невидимыми
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Properties/AssemblyInfo.cs]: - [x] no russian :exclamation:

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+            }
+            catch (Exception ex)
+            {
+                const string Message = "Some errore appears during EntityFramework command execution.";
+                Logger.ErrorException(Message, ex);
+                throw new RepositoryException(Message, ex);
+            }
+        }
+
+        private void Dispose(bool item)
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [x] don't break template rules. is "disposing" instead of "item"

Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [x] finish Disposable pattern implementation

File: Common/Common.DataBaseRepositories/Repository.cs (hunk 7 above)
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [x] Think on exception handling pattern. and use it in each method

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+namespace DataBaseRepositories
+{
+    /// <summary>
+    /// Provides base methods for working with entities.
+    /// </summary>
+    /// <typeparam name="TEntity">The mapped type.</typeparam>
+    /// <typeparam name="TContext">Type of the used DbContext.</typeparam>
+    public class Repository<TEntity, TContext> : IRepository<TEntity>, IDisposable
+        where TEntity : class
+        where TContext : DbContext, new()
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [x] no need in new() restriction

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    {
+        private readonly TContext _context;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Repository{TEntity,TContext}" /> class.
+        /// </summary>
+        /// <param name="context">The DBContext.</param>
+        public Repository(TContext context)
+        {
+            _context = context;
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [ ] check null

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+                Logger.ErrorException(Message, ex);
+                throw new RepositoryException(Message, ex);
+            }
+        }
+
+        private void Dispose(bool item)
+        {
+            if (item)
+            {
+                if (DataContext != null)
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [ ] doesn't need check null

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+                {
+                    DataContext.Dispose();
+                }
+            }
+        }
+
+        ////TODO: Don't remove it, maybe methods will be useful.
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [ ] :fire:

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+        }
+
+        /// <summary>
+        /// Gets all entities with specified type from DB.
+        /// </summary>
+        public IEnumerable<TEntity> GetAll()
+        {
+            try
+            {
+                return DataContext.Set<TEntity>();
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [ ] will it return all at once or iterator?

File: Common/Common.DataBaseRepositories/Repository.cs [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+        /// Deletes user from DB.
+        /// </summary>
+        /// <param name="entity">The object to delete.</param>
+        public void Delete(TEntity entity)
+        {
+            try
+            {
+                var objectSet = ((IObjectContextAdapter)DataContext).ObjectContext.CreateObjectSet<TEntity>();
+                objectSet.Attach(entity);
+                objectSet.DeleteObject(entity);
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [x] Save?

File: Common/Common.DataBaseRepositories/Repository.cs (hunk 7 above)
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/Repository.cs]: - [x] argument null exception everywhere

File: Common/Common.DataBaseRepositories/RepositoryFactory.cs [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        private readonly IList<Type> _registratedEntities = new List<Type>()
+            {
+                typeof(User)
+            };
+
+        /// <summary>
+        /// Creates repositories.
+        /// </summary>
+        /// <typeparam name="TEntity">Type of repository.</typeparam>
+        [SuppressMessage("Microsoft.Reliability", "CA2000:Ликвидировать объекты перед потерей области", Justification = "We don't need it.")]
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/RepositoryFactory.cs]: o_O

File: TActive.sln [hunk 17]
```diff
@@ -1,8 +1,6 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 2013
```
Comment (iivchenko) [inline TActive.sln:3]: - [ ] nope

File: Common/Common.DataBaseRepositories/RepositoryFactory.cs [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        public IRepository<TEntity> CreateRepository<TEntity>() where TEntity : class
+        {
+            var context = new TActiveContext();
+
+            if (_registratedEntities.Contains(typeof(User)))
+            {
+                return new Repository<TEntity, TActiveContext>(context);
+            }
+
+            throw new NotSupportedException(string.Format("In DataBase it doesn't exist table mapped from type {0}", typeof(TEntity)));
```
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/RepositoryFactory.cs]: - [x] grammar

File: Common/Common.DataContracts/User.cs [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+            _profile = profile;
+            _friends = new Friends();
+            Id = Guid.NewGuid();
+        }
+
+        /// <summary>
+        /// Gets users id.
+        /// </summary>
+        [Key]
+        public Guid Id { get; private set; }
```
Comment (iivchenko) [inline Common/Common.DataContracts/User.cs]: - [x] for database performance it is better to user some simple type

File: Common/Common.DataContracts/User.cs [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+// // </copyright>
+// // <author>Ivan Ivchenko</author>
+// // <author>Myroslava Tarcha</author>
+
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+using System.Runtime.Serialization;
+
+namespace Common.DataContracts
```
Comment (iivchenko) [inline Common/Common.DataContracts/User.cs:11]: - [x] to single property style

File: Common/Common.DataContracts/User.cs [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                return _friends;
+            }
+        }
+
+        /// <summary>
+        /// Adds a new contact to users friends.
+        /// </summary>
+        /// <param name="friend">Other user.</param>
+        public void AddFriend(User friend)
```
Comment (iivchenko) [inline Common/Common.DataContracts/User.cs]: - [x] redundant

File: Common/Common.DataContracts/UserProfile.cs [hunk 22]
```diff
@@ -0,0 +1,40 @@
+﻿// // <copyright company="Tarcha & Ivchenko Company">
+// //      Copyright (c) 2015, All Right Reserved
+// // </copyright>
+// // <author>Ivan Ivchenko</author>
+// // <author>Myroslava Tarcha</author>
+
+using System.ComponentModel.DataAnnotations;
```
Comment (iivchenko) [inline Common/Common.DataContracts/UserProfile.cs:8]: - [x] delete unnecessary

File: Common/Common.DataContracts/UserProfile.cs [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+        /// <value>The name of the user.</value>
+        [DataMember]
+        public string AccountName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the hash of entered password.
+        /// </summary>
+        /// <value>The users password hash.</value>
+        [DataMember]
+        public int PasswordHash { get; set; }
```
Comment (iivchenko) [inline Common/Common.DataContracts/UserProfile.cs]: - [x] it is better to use byte[] type and some reliable hash algorithm as MD5

File: Common/Common.DataContracts/UserProfile.cs [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+
+        /// <summary>
+        /// Gets or sets a users age.
+        /// </summary>
+        /// <value>The users age.</value>
+        [DataMember]
+        public int Age { get; set; }
```
Comment (iivchenko) [inline Common/Common.DataContracts/UserProfile.cs]: - [x] why don't you use not negative type

File: Common/Common.Implementation/RepositoryBase.cs [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+using System.Linq.Expressions;
+using NLog;
+
+namespace DataBaseRepositories
+{
+    /// <summary>
+    /// Provides base methods for working with entities.
+    /// </summary>
+    /// <typeparam name="T">EF DbContext.</typeparam>
+    public abstract class RepositoryBase<T> : IDisposable where T : DbContext, new()
```
Comment (iivchenko) [inline Common/Common.Implementation/RepositoryBase.cs]: - [x] do we still need this file?

File: Server/Server.Console/App.config [hunk 26]
```diff
@@ -1,6 +1,22 @@
 ﻿<?xml version="1.0" encoding="utf-8" ?>
 <configuration>
+  <configSections>
+    <!-- For more information on Entity Framework configuration, visit http://go.microsoft.com/fwlink/?LinkID=237468 -->
```
Comment (iivchenko) [inline Server/Server.Console/App.config]: - [x] the link is redundant

File: Server/Server.Console/App.config [hunk 27]
```diff
@@ -1,6 +1,22 @@
...
+    <section name="entityFramework" type="System.Data.Entity.Internal.ConfigFile.EntityFrameworkSection, EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" requirePermission="false" />
+  </configSections>
   <startup>
     <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5" />
   </startup>
+  <entityFramework>
```
Comment (iivchenko) [inline Server/Server.Console/App.config:9]: - [x] is it possible to save configs as a separate file?

File: Server/Server.Console/App.config [hunk 28]
```diff
@@ -1,6 +1,22 @@
...
+    <defaultConnectionFactory type="System.Data.Entity.Infrastructure.SqlConnectionFactory, EntityFramework" />
+    <providers>
+      <provider invariantName="System.Data.SqlClient" type="System.Data.Entity.SqlServer.SqlProviderServices, EntityFramework.SqlServer" />
+    </providers>
+  </entityFramework>
+  <connectionStrings>
+    <add
+      name="MyProductContext"
```
Comment (iivchenko) [inline Server/Server.Console/App.config]: - [x] need better name

File: Server/Server.Console/Program.cs [hunk 29]
```diff
@@ -27,6 +29,8 @@ public static class Program
         /// </summary>
         public static void Main()
         {
+            Database.SetInitializer(new DropCreateDatabaseIfModelChanges<TActiveContext>());
```
Comment (iivchenko) [inline Server/Server.Console/Program.cs:32]: - [x] what for?

File: Common/Common.DataBaseRepositories/IRepository.cs (hunk 3 above)
Comment (mtarcha) [inline Common/Common.DataBaseRepositories/IRepository.cs]: Becouse we can map to DB only reference types.

File: Common/Common.DataBaseRepositories/Repository.cs (hunk 6 above)
Comment (mtarcha) [inline Common/Common.DataBaseRepositories/Repository.cs:7]: I don't see any proplem

File: Common/Common.DataBaseRepositories/IRepository.cs (hunk 3 above)
Comment (iivchenko) [inline Common/Common.DataBaseRepositories/IRepository.cs]: Why?

Comment (mtarcha) [inline Common/Common.DataBaseRepositories/IRepository.cs]: Because

File: Common/Common.DataContracts/UserProfile.cs (hunk 24 above)
Comment (mtarcha) [inline Common/Common.DataContracts/UserProfile.cs]: Type uint is not CLS-compliant

File: Common/Common.DataBaseRepositories/Repository.cs (hunk 9 above)
Comment (mtarcha) [inline Common/Common.DataBaseRepositories/Repository.cs]: what do you mean?

File: Common/Common.DataBaseRepositories/Repository.cs (hunk 14 above)
Comment (mtarcha) [inline Common/Common.DataBaseRepositories/Repository.cs]: what?

File: Common/Common.DataContracts/User.cs (hunk 20 above)
Comment (mtarcha) [inline Common/Common.DataContracts/User.cs:11]: what??

File: Common/Common.DataContracts/UserProfile.cs (hunk 23 above)
Comment (mtarcha) [inline Common/Common.DataContracts/UserProfile.cs]: The implementation of hash algo - the responsibility of the user registration task.

File: Server/Server.Console/Program.cs (hunk 29 above)
Comment (mtarcha) [inline Server/Server.Console/Program.cs:32]: For testing.If you change db schema, and want to check your changes - now you haven't firs delete old db and then create new db - this code will do it for youю

File: Server/Server.Console/App.config (hunk 27 above)
Comment (mtarcha) [inline Server/Server.Console/App.config:9]: Later

**Comment by iivchenko** [comment]
maladez=)

**Pull request merged**